src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/Internal/HttpResponseBodyReader.cs

```diff
@@ -54,90 +53,186 @@ public ValueTask<string> ReadAsync(HttpResponseMessage response, CancellationTok
+        if (readTimeout <= TimeSpan.Zero)
+        {
+            // cancel immediately, async cancel not required in tests
+#pragma warning disable CA1849 // Call async methods when in an async method
+            joinedTokenSource.Cancel();
+#pragma warning restore CA1849 // Call async methods when in an async method
+        }
 
-        var bufferWriter = _bufferWriterPool.Get();
-        var memory = bufferWriter.GetMemory(readSizeLimit).Slice(0, readSizeLimit);
-#if !NETCOREAPP3_1_OR_GREATER
-        byte[] buffer = memory.ToArray();
-#endif
         try
         {
-#if NETCOREAPP3_1_OR_GREATER
-            var charsWritten = await streamToReadFrom.ReadAsync(memory, cancellationToken).ConfigureAwait(false);
-            bufferWriter.Advance(charsWritten);
-            return Encoding.UTF8.GetString(memory.Slice(0, charsWritten).Span);
+            return await ReadFromStreamAsync(response, readSizeLimit, joinedTokenSource.Token).ConfigureAwait(false);
+        }
+
+        // when readTimeout occurred: joined token source is cancelled and cancellationToken is not
+        catch (OperationCanceledException) when (joinedTokenSource.IsCancellationRequested && !cancellationToken.IsCancellationRequested)
+        {
+            return Constants.ReadCancelledByTimeout;
+        }
+    }
+
+    private static async ValueTask<string> ReadFromStreamAsync(HttpResponseMessage response, int readSizeLimit, CancellationToken cancellationToken)
+    {
+#if NET6_0_OR_GREATER
+        Stream streamToReadFrom = await response.Content.ReadAsStreamAsync(cancellationToken).ConfigureAwait(false);
 #else
-            var charsWritten = await streamToReadFrom.ReadAsync(buffer, 0, readSizeLimit, cancellationToken).ConfigureAwait(false);
-            bufferWriter.Advance(charsWritten);
-            return Encoding.UTF8.GetString(buffer.AsMemory(0, charsWritten).ToArray());
+        Stream streamToReadFrom = await response.Content.ReadAsStreamAsync().WaitAsync(cancellationToken).ConfigureAwait(false);
 #endif
+
+        var pipe = new Pipe();
+
+        string bufferedString = await BufferStreamAndWriteToPipeAsync(streamToReadFrom, pipe.Writer, readSizeLimit, cancellationToken).ConfigureAwait(false);
+
+        // if stream is seekable we can just rewind it and return the buffered string
+        if (streamToReadFrom.CanSeek)
+        {
+            streamToReadFrom.Seek(0, SeekOrigin.Begin);
+
+            await pipe.Reader.CompleteAsync().ConfigureAwait(false);
+
+            return bufferedString;
         }
-        finally
+
+        // if stream is not seekable we need to write the rest of the stream to the pipe
+        // and create a new response content with the pipe reader as stream
+        _ = Task.Run(async () =>
         {
-            if (streamToReadFrom.CanSeek)
+            await WriteStreamToPipeAsync(streamToReadFrom, pipe.Writer, cancellationToken).ConfigureAwait(false);
+        }, CancellationToken.None);
```

Comment:
Is this preferred over reusing `cancellationToken`?

Reply:
I originally had it passed both to the inner task to write the stream to pipe, and the background task. My thoughts here is the inner task gets cancelled the thread will terminate anyway. I may be wrong about this. What do we prefer here?

Reply:
I'm not sure, we're using it in [`TcpEndpointProbesService`](https://github.com/dotnet/extensions/blob/09094aebc27b03eec7656bed9c128c5cc523df39/src/Libraries/Microsoft.Extensions.Diagnostics.Probes/TcpEndpointProbesService.cs#L49) like this, but not in [FunctionInvokingChatClient ](https://github.com/dotnet/extensions/blob/09094aebc27b03eec7656bed9c128c5cc523df39/src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs#L441).

After doing some reading, If believe it doesn't really matter in our scenario, given we are using it for the internal call and we don't have [children tasks inside of this `Task.Run()`](https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/how-to-cancel-a-task-and-its-children#example), and we're relying on the status of the cancelled task (`cancelled` vs `faulted`).

There seem to be opinions both for and [against](https://blog.stephencleary.com/2022/02/cancellation-1-overview.html) using it, so your call.

Reply:
> The CancellationToken passed to Task.Run just cancels the scheduling of the delegate to the thread pool; once that delegate starts running (which happens pretty much immediately), that cancellation token is ignored.

To keep semantics, I'm leaving it as-is. That way we are clearly showing that the cancellation should happen for the delegate itself. If we pass it to `Task.Run` it would only prevent the scheduling of the task if it were already cancelled.